spec/p2p/v0.34/README.md

```diff
@@ -0,0 +1,140 @@
+
+Once a peer is successfully dialed, we ask it to provide us with more peers.
+
+#### *Dialing failed*
+
+
+### *Removing peers*
+
+Each reactor can call the `StopPeerForError` method of the switch with the ID of the peer that needs to be removed. Then the switch handles stopping the peer (closing the connection to it), and calls the `RemovePeer` method of all reactors registered to the switch.
+
+## Peer ranking
+
+In v0.34 there is no explicit ranking of peers. When choosing peers to dial, there is slight bias towards new and vetted peers. The amount of bias is higher when there are more peers connected to a node. 
+
+In addition to this, peers  can be marked as bad and removed entirely from the potential candidate list. Reactors themselves can also mark peers as bad or good and thus influence the behaviour of the p2p layer. In v0.34 a peer is marked good only from the consensus reactor whenever a peer delivers a correct consensus message (TODO check conditiosn for this). 
+
+A peer is marked as bad in the following cases: 
+- It sends too frequent requests for peer information (`PexRequest` messages)
+- Returns an error when a node requests peer information from it
+- A node is not able to successfully dial a peer after `MaxDialAttempts` and the peer is not persistent (persistent peers are never marked as bad)
+
+The PEX reactor checks periodically whether (TODO what is the exact condition) a peer can be reinstated and removed from the `bad peer` list. However, this is done only if a node does not have sufficient peers. Otehrwise, this list is never revisited. 
+
+## Switch
+
+Every Tendermint reactor regiters itself with the Switch by prividing a Channel ID it is listening to to it. The Switch then forwards messages destined to a particular channel. It is the reactors responisibility to process the messages. 
+
+The switch is implemented as a service and, on start, listens for incoming connections in the background (calling the`acceptRoutine` function). 
+
+Dialing peers, either via the PEX reactor or dialing persistent peers on startup, is done by calling the `DialPeersAsync` routine of the Switch.
+
+When the connection to a peer is established (either by dialing it or accepting an incoming connection from it), the peer address is added to the address book. (The address book is managed by the PEX reactor). 
+
+When a peer is added to the address book, it is marked as connected to and no further new connections are established between the node and this peer. 
+
+The number of peers a node can connect to is set by `MaxNumInboundPeers` and `MaxNumOutboundPeers` respectively. 
+
+## Pex reactor
+
+The PEX reactor is responsible for peer discovery and providing other peers with information about peers a node is already connected to. The PEX reactor receive routine listens to two types of messages: `PexRequest` and `PexAddress`. 
```

Comment:
> providing other peers with information about peers a node is already connected to

Known peers, or connected peers? In the second case, the provided information is pretty restricted, isn't it?

Reply:
It selects addresses from the address book by calling [`GetSelectionWithBias`](https://github.com/tendermint/tendermint/blob/46badfabd9d5491c78283a0ecdeb695e21785508/p2p/pex/pex_reactor.go#L269).